### ISSUE TITLE
Using json_serializable  v5 causes invalid_annotation_target warnings

### ISSUE THREAD

**Issue by rrousselGit** [rrousselGit/freezed#488]
Example:

```dart
@freezed
class Example with _$Example {
  factory Example(
    @JsonKey() // invalid_annotation_target
    String value,
  ) = _Example
}
```

**Comment by rrousselGit** [comment]
For now my recommendation would be to disable this warning. Such as with:

```yaml
# analysis_options.yaml
analyzer:
  exclude:
    - "**/*.g.dart"
    - "**/*.freezed.dart"
  errors:
    invalid_annotation_target: ignore
```

**Comment by rrousselGit** [comment]
One fix would be to introduce custom annotations  mimicking the behaviour of Freezed annotations, like `@FreezedKey` instead of `@JsonKey`

**Comment by mrverdant13** [comment]
Another workaround would be to add the `// ignore: invalid_annotation_target` comment immediately above the highlighted line. This way, other misplaced annotations are not ignored.

**Comment by rrousselGit** [comment]
Coming back to this, I'll probably end-up not relying on json_serializable in the future and instead manually take care of serialization

**Comment by Vedsaga** [comment]
Not sure, if this error is related to this.
![image](https://user-images.githubusercontent.com/59083528/170017071-6c0da5d2-855a-4cec-b9e6-4456da3a1da3.png)

I am getting below error. 

```
[+10559 ms] [+10435 ms] lib/modules/login/bloc/login_bloc.freezed.dart(150,4): error GD6C8354B: 'JsonKey' is imported from both
'package:drift/src/dsl/dsl.dart' and 'package:json_annotation/src/json_key.dart'.
[C:\bitecope\salesman\build\windows\flutter\flutter_assemble.vcxproj]
```

if required I could share the full log... Due to it. I am stuck and can not run the `flutter run` command... any help on how to solve this temporarily?

**Comment by rrousselGit** [comment]
No this is unrelated. It sounds like you are using drift, which also exports a JsonKey class (which is not from json_serializable). So you have a conflict

**Comment by Vedsaga** [comment]
> conflict

Thank you for quick response.
umm, I am using `drift` ... Not sure, what to do fix this.

**Comment by Vedsaga** [comment]
also, getting this error
![image](https://user-images.githubusercontent.com/59083528/170027053-e891be63-1343-4a99-b168-979319e0b7f3.png)

```
[   +1 ms]   [+10435 ms] lib/modules/login/bloc/login_bloc.freezed.dart(150,4): error GD6C8354B: 'JsonKey' is imported from both
'package:drift/src/dsl/dsl.dart' and 'package:json_annotation/src/json_key.dart'.
```

I am not able to do `flutter run`. Due to which I can't make any progress :(

**Comment by SunlightBro** [comment]
@Vedsaga Try importing drift always with alias like so:
And then always use classes from drift like so:
```dart
import 'package:drift/drift.dart' as drift;

@drift.DriftDatabase(tables: [...], views: [...])
class Database extends _$Database {...}
```
Or hide JsonKey from freezed (json_serializable) if not needed,
```dart
import 'package:freezed_annotation/freezed_annotation.dart' hide JsonKey;
```

**Comment by Vedsaga** [comment]
```dart
import 'package:freezed_annotation/freezed_annotation.dart' hide JsonKey;
```

Thank you so much this solved the issue :)

**Comment by berkaykurkcu** [comment]
> Coming back to this, I'll probably end-up not relying on json_serializable in the future and instead manually take care of serialization

Did I miss any update on this?  @rrousselGit

**Comment by rrousselGit** [comment]
Freezed 3.0 supports normal classes ; which don't suffer from this issue

```dart
@JsonSerializable()
@freezed
class Foo with _$Foo {
  Foo(this.name);
  factory Foo.fromJson(Map<String, Object?> json) => _$FooFromJson(json);
 
  final String name;
}
```

You can also extract individual union cases:

```dart
@freezed
class Union with _$Union {
  factory Union.foo(String name) = Foo;
  factory Union.bar(String name) = Bar;
}

@JsonSerializable()
@freezed
class Foo with _$Foo implements Union {
  Foo(this.name);
  factory Foo.fromJson(Map<String, Object?> json) => _$FooFromJson(json);
 
  final String name;
}
```

So, although more verbose, you can solve the issue.

**Comment by rrousselGit** [comment]
3.0 solves this, so I'll close it.

TL;DR, use a "normal" class https://github.com/rrousselGit/freezed/issues/488#issuecomment-2740682136

**Comment by cbreezier** [comment]
@rrousselGit I'm not sure this is really a valid solution. The freezed documentation still suggests using factory constructors. There's no mention of this issue in the documentation.

I think it's a pretty big limitation, given:
1. How useful and common it is to use `@JsonKey` annotations
2. How much boilerplate is saved by using factory constructors, and how prominently this method is promoted in the freezed documentation

I'd be happy with a `@FreezedKey` replacement.

@mrverdant13 adding the suppression to the exact line doesn't actually work, because that suppression gets copied into the `*.freezed.dart` file and you end up with this warning instead:

`warning: The diagnostic 'invalid_annotation_target' doesn't need to be ignored here because it's already being ignored.`
